Avoiding forced all fields in a contained belongsToMany, fixes #7913

lorenzo opened the pull request:
Also fixes https://github.com/cakephp/cakephp/issues/7916

lorenzo committed:
Avoiding forced all fields in a contained belongsToMany, fixes #7913

lorenzo committed:
Removing test to satisfy travis for now

This particular case is also tested from multiple different angles anyways

lorenzo committed:
Speeding up travis a bit

lorenzo committed:
Qoutes in yaml file were actually important

lorenzo committed:
Adding composer cache for travis

markstory commented on .travis.yml:
```diff
@@ -43,6 +48,7 @@ matrix:
 
 before_script:
   - composer self-update
+  - if [ -n "$GH_TOKEN" ]; then composer config github-oauth.github.com ${GH_TOKEN}; fi;
```
What is this for?

lorenzo replied:
Makes composer I'm Travis use a github oauth token, so that repos can be downloaded and not cloned. It speeds up a bit the builds

ADmad replied:
:clap:

markstory replied:
Sounds awesome!

markstory commented on src/ORM/Association/BelongsToMany.php:
```diff
@@ -1095,11 +1095,23 @@ protected function _collectJointEntities($sourceEntity, $targetEntities)
+        $query->addDefaultTypes($assoc->target());
+
+        if ($queryBuilder) {
+            $query = $queryBuilder($query);
```
Was the queryBuilder never used before?

lorenzo replied:
It was, at the end of the parent's eagerLoad() method. That was too late for limiting the fields, but also the right moment, for the query already had the types information for its associations.

I had to move it here, where it is called before the association is attached, in order to limit the fields; but then had to manually set the types for the associated table myself.

markstory commented on tests/TestCase/Database/QueryTest.php:
```diff
@@ -336,17 +336,6 @@ public function testSelectInnerJoin()
-            ->execute();
-        $this->assertEquals(
-            ['title' => 'First Article', 'name' => 'Second Comment for First Article'],
-            $result->fetch('assoc')
```
Good by flakey test.

markstory commented on tests/TestCase/ORM/QueryRegressionTest.php:
```diff
@@ -1236,4 +1236,26 @@ public function testCountWithComplexOrderBy()
+            ->first();
+
+        $this->assertNotEmpty($result->tags[0]->id);
+        $this->assertEmpty($result->tags[0]->name);
```
The BelongsToMany test case has fixtures now. This could go there instead.

bravo-kernel commented:
Thanks :thumbsup:

lorenzo committed:
Moving test case to another file

lorenzo committed:
Trying to speed up builds a bit more by disabling xdebug

lorenzo committed:
Trying to fix some tests in appveyor

lorenzo committed:
Repeating previous fix for HasMany associations.

This makes appveyor happy

lorenzo committed:
Caching composer packages in appveyor

Some debugging code for appveyor

More debugging for appveyor

Better cache folder for appveyor

More debugging code for appveyor

lorenzo committed:
Trying to figure out what sql server is doing

lorenzo committed:
Skipping test in SQLServer while we find an appropriate fix

This test is at least an obscure feature of the ORM, and fixing it
properly requires a fair bit of changes. Skipping for now

lorenzo committed:
Restoring appveyor.yml

lorenzo committed:
Still trying to figure out the right composer cache folder for appveyor

lorenzo committed:
Fixing appveyor.yaml

lorenzo commented:
I fixed all the tests in appveyor as well. Had to skip one of them since fixing it properly requires some big changes.

markstory commented:
@lorenzo :clap: :clap: :clap:
